packages/@sanity/codegen/src/typescript/schemaTypeGenerator.ts

```diff
@@ -0,0 +1,381 @@
+        }
+      }
+    }
+
+    if (typeNode.dereferencesTo) {
+      const derefType = Object.assign(
+        t.tsPropertySignature(
+          SchemaTypeGenerator.InternalReferenceSymbol,
+          t.tsTypeAnnotation(t.tsLiteralType(t.stringLiteral(typeNode.dereferencesTo))),
+        ),
+        {computed: true, optional: true},
+      )
+      props.push(derefType)
+    }
+
+    return t.tsTypeLiteral(props)
+  }
+
+  private generateInlineTsType(typeNode: InlineTypeNode): t.TSType {
+    const identifier = this.getIdentifier(typeNode.name)
+    if (!identifier) {
+      // Not found in schema, return unknown type
+      return t.addComment(
+        t.tsUnknownKeyword(),
+        'trailing',
+        ` Unable to locate the referenced type "${typeNode.name}" in schema "${this.schemaId}"`,
+        true,
+      )
+    }
+
+    return t.tsTypeReference(identifier)
+  }
+
+  // Helper function used to generate TS types for union type nodes.
+  private generateUnionTsType(typeNode: UnionTypeNode): t.TSType {
+    if (typeNode.of.length === 0) return t.tsNeverKeyword()
+    if (typeNode.of.length === 1) return this.generateTsType(typeNode.of[0])
+    return t.tsUnionType(typeNode.of.map((node) => this.generateTsType(node)))
+  }
+
+  // Helper function used to generate TS types for document type nodes.
+  private generateDocumentTsType(document: DocumentSchemaType): t.TSType {
+    const props = Object.entries(document.attributes).map(([key, node]) =>
+      this.generateTsObjectProperty(key, node),
+    )
+
+    return t.tsTypeLiteral(props)
+  }
+
+  getTypeNames(): string[] {
+    return Array.from(this.typeNames)
+  }
+
+  getTsType(typeName: string): t.TSType | undefined {
+    return this.tsTypes.get(typeName)
+  }
+
+  getIdentifier(typeName: string): t.Identifier | undefined {
+    return this.identifiers.get(typeName)
+  }
+
+  evaluateQuery(query: string): {tsType: t.TSType; stats: TypeEvaluationStats} {
+    const ast = safeParseQuery(query)
+    const typeNode = typeEvaluate(ast, this.schema)
+    const tsType = this.generateTsType(typeNode)
+    const stats = walkAndCountQueryTypeNodeStats(typeNode)
+    return {tsType, stats}
+  }
+
+  evaluateProjection(projection: string): {tsType: t.TSType; stats: TypeEvaluationStats} | null {
+    const results: {tsType: t.TSType; stats: TypeEvaluationStats}[] = []
+
+    for (const schemaType of this.schema) {
+      // only document types are supported for projections
+      if (schemaType.type !== 'document') continue
+
+      const projectionAst = safeParseQuery(projection)
+      if (projectionAst.type !== 'Object') {
+        throw new Error(
+          `Invalid projection syntax: Projections must be enclosed in curly braces, (e.g., "{_id, title}"). Received: "${projection}"`,
+        )
+      }
+
+      const ast = safeParseQuery(`* [_type==${JSON.stringify(schemaType.name)}] ${projection} [0]`)
+      const result = typeEvaluate(ast, this.schema)
```

Comment:
Also want to double check the assumption here - we evaluate projection on _every_ schema, and we are fine with this being slow and generating a lot of unused types that are then reduced during type checking? 😅